netket/graph/lattice.py

```diff
@@ -104,323 +107,349 @@ def get_edges(atoms_positions, cutoff, distance_atol=cutoff_tol):
     return pairs
 
 
-def create_points(basis_vectors, extent, atom_coords, pbc):
+@dataclass
+class LatticeSite:
+    id: int
+    """Integer ID of this site"""
+    position: Array
+    """Real-space position of this site"""
+    cell_coord: Array
+    """Cell coordinates of this site"""
+    inside: bool
+    """TODO: When exactly is this needed?"""
+
+    def __repr__(self):
+        s = ", ".join(map(str, (self.id, self.cell_coord)))
+        return f"LatticeSite({s})"
+
+
+def create_sites(primitives, extent, apositions, pbc):
     shell_vec = _np.zeros(extent.size, dtype=int)
     shift_vec = _np.zeros(extent.size, dtype=int)
     # note: by modifying these, the number of shells can be tuned.
     shell_vec[pbc] = 2
     shift_vec[pbc] = 1
```

Comment:
What do these things do? It seems they make sites non-unique

Reply:
I'm not sure what is meant by "shells" here, so I am unsure.

Reply:
Ah, I think I get it. The nearest-neighbour search is made to work with PBC with padding the PBC dimensions with a few extra unit cells (called here "shells"). Then in `create_edges` these are collapsed on each other, and the "sorting" is needed to only keep these unique versions. Arrrgh

Reply:
OK, it seems we have to live with this — I wouldn't attempt to change anything about `create_edges`. 
It seems the "sorting" returns edges in their expected order, though I couldn't argue why. Maybe we can add an explicit `lexsort` somewhere if we want to be sure, but I can live without it.

Reply:
So the issues is indeed this: `create_sites` creates extraneous sites (in connection with these shell and shift vecs) that are later removed by the "node renaming" logic further down. Even without those, however, start to see issues with this test:
```
assert graph.n_nodes == unit_cells[i] * atoms_per_unit_cell[i]
```
So something is not consistent here.

Reply:
Going forward, it would be nice not to have to use this logic when creating a `Lattice`. I, for one, would love to be able to use second-neighbour Hamiltonians, or colour nearest-neighbour edges (e.g. for the Kitaev model). But let's leave that for later, I really want to put out a PR about symmetries already :smiley:

Reply:
Not sure how much sense it makes to dig into that here more.

For now, we have tests for the ordering and basic consistency of the site information and this code produces a results that satisfy those tests. Since you want to refactor that code anyways this seems sufficient to me for the moment.

Reply:
> So the issues is indeed this: `create_sites` creates extraneous sites (in connection with these shell and shift vecs) that are later removed by the "node renaming" logic further down. Even without those, however, I start to see issues with this test once I remove the node renaming logic:
> 
> ```
> assert graph.n_nodes == unit_cells[i] * atoms_per_unit_cell[i]
> ```
> 
> So something is not consistent here in this regard and the original `sites` and `self._sites` indeed differ quite a bit.

You mean this assert sill fails if you turn off PBCs?

Reply:
> Not sure how much sense it makes to further dig into that here.
> 
> For now, we have tests for the ordering and basic consistency of the site information and this code produces a results that satisfy those tests. Since you want to refactor that code anyways this seems sufficient to me for the moment.

Agreed for now. At some later time I'll package all this logic in one function, so more versatile things can be built interchangeably with it. I don't want to understand what this thing is doing :sweat_smile:

Reply:
Probably not, it failed for some of the test cases and most of them do have PBC.

Reply:
Hey, again sorry for joining in late. @attila-i-szabo is right. Shells are just extra cells around the lattice that are used to match nodes of the lattice from one end to the other when PBC conditions are used. The padding has only 1 shell (by this point you know what I mean by shell), but it might have more... why? This has to do with the [minimum image convention](https://en.wikipedia.org/wiki/Periodic_boundary_conditions#Practical_implementation:_continuity_and_the_minimum_image_convention), which was discussed when [building the Lattice  class in Python for the first time](https://github.com/netket/netket/issues/371). Essentially, it seems that for very skewed unit cells it might be possible that more shells are needed.

Reply:
@PhilipVinc recommended me to write some docs on those functions. I thought it was a good practical choice to solve the matching-nodes problem with a KDTree search, as it is quite fast. But if either @femtobit or @attila-i-szabo want to improve on that, I'd be happy to help, as I finally have time to return to this. If the code will be left as is for now, I'll write the docs for those strange looking functions! Let me know what you decide

Reply:
@VolodyaCO As far as I'm concerned, the lattice construction code is fine and, in particular, the kd tree search seems like a very reasonable approach to me.

Actually, it could make sense to store the kdtree with the lattice and use it down the line. In principle, checking whether a lattice site exists at a given point can be done by using the kdtree nearest-neighbor lookup operation and checking whether that distance is below the desired threshold. (Do we actually need the "integer position lookup" for the symmetries in that case?)

So, given the tree, can we implement the position-lookup as
```python
def id_from_position(pos):
    closest_site_pos = self.kdtree.query(pos)
    return np.linalg.norm(pos - closest_site_pos) < threshold
```
That would siginficantly simplify things.

Reply:
The difficulty is of course, that we have to use the KDTree with the additional shells in order to correctly lookup sites over the boundaries which is needed for the symmetries (right?). While the original kdtree used to construct the edges has those shells, we also need to store the mapping `<index in kdetree> -> <our lattice site id>`. However, that mapping is more or less used in your node-renaming code in `__init__` which removes the duplicates, right @VolodyaCO> Maybe we can use that.